### DISCUSSION TITLE
Seeking Optimization for Video Streaming Route in FastAPI - Facing Slowness Issue

### DISCUSSION THREAD

**Discussion by gustavosett** [fastapi/fastapi#10104] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get(
    "/video_feed/{camera_id}",
    responses={200: {"content": {"multipart/x-mixed-replace;boundary=frame": {}}}},
)
async def video_feed(request: Request, camera_id: int):
    # frames generator
    async def stream_frames():
        while CAMERA_MANAGER[camera_id].is_running:
            # read frame from camera's manager
            frame = await CAMERA_MANAGER[camera_id].get_frame()
            # encode frame as jpeg
            ret, buffer = cv2.imencode(".jpg", frame)

            # yield frame in byte format
            yield (
                b"--frame\r\n"
                b"Content-Type: image/jpeg\r\n\r\n" + buffer.tobytes() + b"\r\n"
            )
    # return generator with StreamingResponse object
    return StreamingResponse(
        stream_frames(), media_type="multipart/x-mixed-replace;boundary=frame"
    )
```

## Hi all,

I've been banging my head against the wall trying to set up a live streaming video route with FastAPI(for security cameras). It's working, but I'm getting slow streaming, low fps, and less-than-stellar performance. I've done my best to troubleshoot it, but now I'm turning to you all, hoping someone might have some clever tricks or know of some third-party tools to get me out of this jam.

### Here's what I'm doing so far:

- Camera ID Parameterization: I'm using camera_id to decide which camera to stream from. Simple enough.
- Frame Streaming: I've got a generator function that keeps grabbing frames from the camera, as long as it's running.
- Frame Encoding: Encoding each frame as a JPEG with OpenCV's cv2.imencode. Standard stuff.
- Response Building: Using FastAPI's StreamingResponse, I'm sending the frames with the multipart/x-mixed-replace;boundary=frame media type.

### This all works, but the stream is painfully slow. I've dug into potential bottlenecks, but I'm kind of stuck. Here's what I'm wondering:
1. Encoding Efficiency: Is there a faster way to encode these frames, or maybe some third-party tool that could do it quicker?
2. Concurrency and Parallelism: Can I tweak things to handle multiple streams better, or do something clever with concurrency?
3. Deployment Environments: Should I be looking at server configurations or other techniques to make the streaming go faster?
4. Third-Party Solutions: Does anyone know of libraries or tools designed to handle video streaming more effectively in FastAPI?

### I'm not married to my current approach, so if there's a way to really ramp up the streaming speed, I'm all ears. I'll include the code snippet below.

Thank you all for any help you can provide. This one's got me stumped!

> P.S. Feel free to respond with any wild ideas, no matter how out there they might be. I'm ready to try anything!

### Operating System

Windows

### FastAPI Version

0.100.0

### Pydantic Version

2.1.1

### Python Version

Python 3.11.3

**Comment by romabozhanovgithub** [comment]
Hey,

1. #### It's possible that the bottleneck is in the encoding process.
You could try using a different encoding format, such as `H.264`, which is more efficient than `JPEG ` for video streaming. You could also experiment with different encoding parameters to find the optimal balance between quality and performance.

Here's an example of how you could optimize it by using the WebP format:
```python
@app.get(
    "/video_feed/{camera_id}",
    responses={200: {"content": {"multipart/x-mixed-replace;boundary=frame": {}}}},
)
async def video_feed(request: Request, camera_id: int):
    # frames generator
    async def stream_frames():
        while CAMERA_MANAGER[camera_id].is_running:
            # read frame from camera's manager
            frame = await CAMERA_MANAGER[camera_id].get_frame()
            # encode frame as WebP
            ret, buffer = cv2.imencode(".webp", frame)

            # yield frame in byte format
            yield (
                b"--frame\r\n"
                b"Content-Type: image/webp\r\n\r\n" + buffer.tobytes() + b"\r\n"
            )
    # return generator with StreamingResponse object
    return StreamingResponse(
        stream_frames(), media_type="multipart/x-mixed-replace;boundary=frame"
    )
```

2. #### Also you can try to reduce the frame rate:
If the video is being streamed at a high frame rate, reducing the frame rate could reduce the amount of data being sent and improve performance.

3. #### You can use a CDN:
If the video is being streamed to a large number of users, using a Content Delivery Network (CDN) could improve performance by caching the video content closer to the users.

4. #### Third-Party Solutions:
You could try using the `aiortc` library for real-time communication between browsers and servers, or the `ffmpeg` library for video encoding and decoding.
Here is an example of using `aiortc`:
```python
import asyncio
from fastapi import FastAPI
from fastapi.responses import StreamingResponse
from aiortc import MediaStreamTrack, RTCPeerConnection, RTCSessionDescription

app = FastAPI()


class VideoStreamTrack(MediaStreamTrack):
    def __init__(self):
        super().__init__()        
        self.cap = cv2.VideoCapture(0)

    async def recv(self):
        ret, frame = self.cap.read()
        if not ret:
            return
        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        return frame


@app.get("/video_feed")
async def video_feed():
    pc = RTCPeerConnection()
    video_track = VideoStreamTrack()
    pc.addTrack(video_track)

    offer = await pc.createOffer()
    await pc.setLocalDescription(offer)

    async def stream_video():
        while True:
            frame = await video_track.recv()
            if frame is None:
                break
            ret, buffer = cv2.imencode(".jpg", frame)
            yield (
                b"--frame\r\n"
                b"Content-Type: image/jpeg\r\n\r\n" + buffer.tobytes() + b"\r\n"
            )

    return StreamingResponse(
        stream_video(),
        media_type="multipart/x-mixed-replace;boundary=frame"
    )
```

**Comment by MubashirWaheed** [comment]
@gustavosett how did you solve the issue? I am also facing the same issue

**Reply by gustavosett** [reply]
There is nothing we can do. :P

**Reply by MubashirWaheed** [reply]
no  :(

**Comment by bassamalasadi** [comment, marked as answer]
Have you tried to use **PyTurboJPEG**?
`jpeg.encode(frame, quality=60)`
**Benchmark**
For Windows

Function | Wall-clock time
-- | --
cv2.imdecode() | 0.358 sec
TurboJPEG.decode() | 0.135 sec
cv2.imencode() | 0.581 sec
TurboJPEG.encode() | 0.140 sec

**Comment by Praveenstein** [comment]
Hey @gustavosett , I recently faced similar issues with video streaming in FastAPI and managed to optimize it using a different approach. In my case, I switched to a more efficient method using WebCodecs and Web Workers for frame decoding and streaming. This drastically improved performance, reduced latency, and minimized bandwidth consumption compared to traditional methods like JPEG encoding.

I wrote a detailed blog about how I approached low-latency video streaming with FastAPI, WebCodecs, and Web Workers, which you might find useful. It covers everything from encoding to streaming and includes tips for improving concurrency and overall performance.

You can check it out here: [Low Latency Video Streaming Application with FastApi](https://medium.com/@praveen06061995/low-latency-video-streaming-application-with-fastapi-4c69f5124fda)
[Github Repo](https://github.com/Praveenstein/low-latency-streamer)

It might offer some insights on optimizing frame encoding and stream handling that could help with your security camera stream performance.

Hope it helps! Let me know if you have any questions. 😃
